### PR TITLE
Check for trailing garbage inside json

### PR DESCRIPTION
**Pull request by tmegow**
A stab at addressing open issue #73

### CODE REVIEW

**Review by LB-- — CHANGES_REQUESTED** [review]
There are too many formatting/spacing changes that make it difficult to see the actual changed code. Would you be willing to try again without reformatting the source file?

**Comment by DimitriPapadopoulos** [comment]
@tmegow Could you perhaps discard the [indentation](https://github.com/json-parser/json-parser/pull/80/commits/32524f513ca88e4467e71cf221a6c181c4015588) and [formatting](https://github.com/json-parser/json-parser/pull/80/commits/86d646c1290528e51a22bfd0857d7cc80898a719) commits? Then, commit [Allow for empty objects and empty arrays](https://github.com/json-parser/json-parser/pull/80/commits/d63b9d80a6053c66241468322ee314d6db303292) must be reworked to only bring essential changes.

I'm not saying the current format is perfect, but reformatting everything while adding actual changes is never an option.

File: json.h [hunk 1]
```diff
@@ -266,6 +266,7 @@ json_value * json_parse_ex (json_settings * settings,
 
 void json_value_free (json_value *);
 
+int trailing_garbage (const json_char *);
```
Comment (DimitriPapadopoulos) [inline json.h:269]: Does this function need to be declared in `json.h`? I don't think so, it should be a static function of `json.c`.

File: json.c [hunk 2]
```diff
@@ -202,810 +216,914 @@ static int new_value (json_state * state,
...
+    switch (*ptr)
+    {
+        case '}':
+        case '{':
+        case ']':
+        case '[':
+        case '"':
+            return 0;
+
+        case 'e':
```
Comment (DimitriPapadopoulos) [inline json.c:1090]: It looks like you're looking for `"false"` Perhaps a `strncmp()` would be more readable. At least add a comment, like `/* skipping "false" */` or something similar.

File: json.c [hunk 3]
```diff
@@ -202,810 +216,914 @@ static int new_value (json_state * state,
...
+            marker = *(--ptr);
+            if (marker == 's')
+            {
+                if (*(--ptr) == 'l' && *(--ptr) == 'a' && *(--ptr) == 'f')
```
Comment (DimitriPapadopoulos) [inline json.c:1094]: Here you decrement `--ptr` multiple times. Are there cases where you could decrement pas the beginning of the buffer?

File: json.c [hunk 4]
```diff
@@ -202,810 +216,914 @@ static int new_value (json_state * state,
...
+                {
+                    return 0;
+                }
+            }
+            if (marker == 'u')
```
Comment (DimitriPapadopoulos) [inline json.c:1099]: It looks like you're looking for "true" Perhaps a strncmp() would be more readable. At least add a comment, like /* skipping "true" */ or something similar.

File: json.c [hunk 5]
```diff
@@ -202,810 +216,914 @@ static int new_value (json_state * state,
...
+            {
+                if (*(--ptr) == 'r' && *(--ptr) == 't')
```
Comment (DimitriPapadopoulos) [inline json.c:1101]: Same comment about multiple `--ptr`.

File: json.c [hunk 6]
```diff
@@ -202,810 +216,914 @@ static int new_value (json_state * state,
...
+                {
+                    return 0;
+                }
+            }
+
+            return 1;
+
+        case 'l':
```
Comment (DimitriPapadopoulos) [inline json.c:1109]: It looks like you're looking for `"null"` Perhaps a `strncmp()` would be more readable. At least add a comment, like `/* skipping "null" */` or something similar.

File: json.c [hunk 7]
```diff
@@ -202,810 +216,914 @@ static int new_value (json_state * state,
...
+            if (*(--ptr) == 'l' && *(--ptr) == 'u' && *(--ptr) == 'n')
```
Comment (DimitriPapadopoulos) [inline json.c:1110]: Same comment about multiple `--ptr`.

File: json.c (hunk 2 above)
Comment (DimitriPapadopoulos) [reply json.c:1090]: But then would `strncmp()` work for all possible types of `json_char`?

**Pull request closed without merging**